File: design-pattern/command-pattern/command.cpp
#include <iostream>

using namespace std;

// command interface
class command
{
public:
    virtual void execute() = 0; // Pure virtual function
    virtual void undo() = 0;    // Pure virtual function
    virtual ~command() {}       // Virtual destructor
};

// receiver classes
class Light
{
public:
    void on()
    {
        cout << "Light is ON" << endl;
    }

    void off()
    {
        cout << "Light is OFF" << endl;
    }
};

class Fan
{
public:
    void on()
    {
        cout << "Fan is ON" << endl;
    }

    void off()
    {
        cout << "Fan is OFF" << endl;
    }
};

// concrete command classes

class LightOnCommand : public command
{
private:
    Light *light;

public:
    LightOnCommand(Light *l) : light(l) {};

    void execute() override
    {
        light->on();
    }

    void undo() override
    {
        light->off();
    }
};

class FanOnCommand : public command
{
private:
    Fan *fan;

public:
    FanOnCommand(Fan *f) : fan(f) {};

    void execute() override
    {
        fan->on();
    }

    void undo() override
    {
        fan->off();
    }
};

// invoker class Remote Controller with static 2d array of 6 button

class RemoteController
{
private:
    static const int numButtons = 6;
    command *buttons[numButtons];   // 2D array for commands and undo command
    bool buttonPressed[numButtons]; // Track button pressed state
public:
    RemoteController()
    {
        for (int i = 0; i < numButtons; i++)
        {
            buttons[i] = nullptr;
            buttonPressed[i] = false;
        }
    }

    void setCommand(int index, command *cmd)
    {
        if (index >= 0 && index < numButtons)
        {
            if (buttons[index] != nullptr)
            {
                delete buttons[index]; // Clean up previous command
                buttons[index] = cmd;
                buttonPressed[index] = false; // Reset button state
            }
        };
    };

    void pressButton(int index)
    {
        if (index >= 0 && index < numButtons && buttons[index] != nullptr)
        {
            if (buttonPressed[index] == false)
            {
                buttons[index]->execute(); // Mark button as pressed
            }
            else
            {
                buttons[index]->undo(); // Undo the command if already pressed
            }

            buttonPressed[index] = !buttonPressed[index]; // Toggle button state
        }
        else
        {
            cout << "Invalid button index or command not set." << endl;
        }
    }

    ~RemoteController()
    {
        for (int i = 0; i < numButtons; i++)
        {
            delete buttons[i]; // Clean up all commands
        }
    };
};

int main(void)
{
    Light *livingLight = new Light();
    Fan *ceilingFan = new Fan();

    RemoteController *remote = new RemoteController();

    remote->setCommand(0, new LightOnCommand(livingLight));

    remote->setCommand(1, new FanOnCommand(ceilingFan));

    // Pressing button 0 to turn on the light
    remote->pressButton(0); // Output: Light is ON
    // Pressing button 0 again to turn off the light
    remote->pressButton(0); // Output: Light is OFF
}

File: design-pattern/command-pattern/command.ts

interface command {
    run(): void;
    undo(): void;
}

class Light {
    on() {
        console.log("light is on");
    }

    off() {
        console.log("light is off");
    }
}

class Fan {
    on() {
        console.log("fan is on");
    }

    off() {
        console.log("fan is off");
    }
}

class LightCommand implements command {
    private light: Light;

    constructor(light: Light) {
        this.light = light;
    }

    run(): void {
        this.light.on();
    }

    undo(): void {
        this.light.off();
    }
}

class FanCommand implements command {
    private fan: Fan;

    constructor(fan: Fan) {
        this.fan = fan;
    }

    run(): void {
        this.fan.on();
    }

    undo(): void {
        this.fan.off();
    }
}

class RemoteController {
    private static readonly numButton: number = 6;
    private buttons: (command | null)[] = new Array(RemoteController.numButton);

    private buttonPressed: boolean[] = new Array(RemoteController.numButton);

    constructor() {
        for (let i = 0; i < RemoteController.numButton; i++) {
            this.buttons[i] = null;
            this.buttonPressed[i] = false;
        }
    }

    setCommand(index: number, cmd: command): void {
        if (index >= 0 && index < RemoteController.numButton) {
            this.buttons[index] = cmd;
            this.buttonPressed[index] = false;
        } else {
            console.log("Invalid button index");
        }
    }

    pressButton(index: number): void {
        if (
            index >= 0 &&
            index < RemoteController.numButton &&
            this.buttons[index]
        ) {
            if (!this.buttonPressed[index]) {
                this.buttons[index]!.run();
            } else {
                this.buttons[index]!.undo();
            }

            this.buttonPressed[index] = !this.buttonPressed[index];
        } else {
            console.log("Invalid button index or command not set");
        }
    }
}


const livingLight = new Light();

const remonte = new RemoteController();

remonte.setCommand(0, new LightCommand(livingLight));

remonte.pressButton(0);
remonte.pressButton(0);